Add own-CA instructions to self-hosted DB guides

ptgott opened the pull request:
Closes #14728

In guides where we show how to configure Teleport to protect a self-hosted database that uses mTLS to communicate with the Database Service, show how to configure the Database Service to trust a custom CA.

This includes the following self-hosted database guides:

- Postgres
- Vitess
- MySQL
- Elasticsearch
- ClickHouse
- Standalone Redis
- Standalone MongoDB
- Cassandra/Scylla
- Oracle

This change does not touch guides that:
- Already include instructions to configure the Database Service to trust a custom CA via the `ca_cert_file` config field, e.g., the CockroachDB and Redis cluster guides.
- Use a different authentication mechanism, e.g., the SQL Server PKINIT guide.

ptgott committed:
Add own-CA instructions to self-hosted DB guides

Closes #14728

In guides where we show how to configure Teleport to protect a
self-hosted database that uses mTLS to communicate with the Database
Service, show how to configure the Database Service to trust a custom
CA.

This includes the following self-hosted database guides:

- Postgres
- Vitess
- MySQL
- Elasticsearch
- ClickHouse
- Standalone Redis
- Standalone MongoDB
- Cassandra/Scylla
- Oracle

This change does not touch guides that:
- Already include instructions to configure the Database Service to
  trust a custom CA via the `ca_cert_file` config field, e.g., the
  CockroachDB and Redis cluster guides.
- Use a different authentication mechanism, e.g., the SQL Server PKINIT
  guide.

greedy52 commented:
@GavinFrazar @Tener could you take a look at the databases that you are familiar? Thanks!

greedy52 commented on docs/pages/database-access/guides/sql-server-ad-pkinit.mdx:
````diff
@@ -73,7 +73,7 @@ You will need to repeat these steps if you rotate Teleport's database certificat
 
    ```code
-   $ tctl auth crl --type=db_client > db-ca.crl
+   $ tctl auth crl --type=db-client > db-ca.crl
````
I believe `db_client` was correct here. The same goes for `tctl auth rotate`. Only `tctl auto export` uses `db-client` as `type`. @GavinFrazar could you confirm?

GavinFrazar replied:
unfortunately for our UX, that's right: `tctl auth export` uses hyphens, the others do not.

Tener commented on docs/pages/database-access/guides/clickhouse-self-hosted.mdx:
````diff
@@ -121,95 +111,149 @@ you're running multiple services add `--output=stdout` to print the config in
 
 ```code
-$ teleport db configure create \
+$ sudo teleport db configure create \
````
Why `sudo`?

ptgott replied:
I don't think there was a conscious reason. My guess is I copied this command from another guide that used it, `sudo` included.

GavinFrazar replied:
```bash
$ teleport db configure create -o file
ERROR: could not write to config file, missing sudo?
	open /etc/teleport.yaml: permission denied
```

this must be why 😅 please re-add

Tener reviewed (commented):
Did you test each particular engine?

GavinFrazar commented on docs/pages/database-access/guides/postgres-self-hosted.mdx:
```diff
@@ -27,6 +27,8 @@ description: How to configure Teleport database access with self-hosted PostgreS
 - Command-line client `psql` installed and added to your system's `PATH` environment variable.
 - A host, e.g., an Amazon EC2 instance, where you will run the Teleport Database
   Service.
+- Optional: a certificate authority that issues certificates and private keys for
+  your self-hosted database.
```
nit: CAs only issue certificates

```suggestion
- Optional: a certificate authority that issues certificates for
  your self-hosted database.
```

GavinFrazar commented on docs/pages/database-access/guides/oracle-self-hosted.mdx:
```diff
@@ -25,6 +25,8 @@ description: How to configure Teleport database access with Oracle.
 - Self-hosted Oracle server instance 18c or later.
 - The `sqlcl` [Oracle client](https://www.oracle.com/pl/database/sqldeveloper/technologies/sqlcl/) installed and added to your system's `PATH` environment variable or any GUI client that supports JDBC
   Oracle thin client.
+- Optional: a certificate authority that issues certificates and private keys for
+  your self-hosted database.
```
nit: CAs only issue certificates

```suggestion
- Optional: a certificate authority that issues certificates for
  your self-hosted database.
```

GavinFrazar commented on docs/pages/database-access/guides/mysql-self-hosted.mdx:
```diff
@@ -26,6 +26,8 @@ description: How to configure Teleport database access with self-hosted MySQL/Ma
 - A self-hosted MySQL or MariaDB instance.
 - A host, e.g., an Amazon EC2 instance, where you will run the Teleport Database
   Service.
+- Optional: a certificate authority that issues certificates and private keys for
+  your self-hosted database.
```
nit: CAs only issue certificates

```suggestion
- Optional: a certificate authority that issues certificates for
  your self-hosted database.
```

GavinFrazar commented on docs/pages/database-access/guides/elastic.mdx:
```diff
@@ -19,6 +19,9 @@ description: How to configure Teleport database access with Elasticsearch.
 
   See [Installation](../../installation.mdx) for details.
 
+- Optional: a certificate authority that issues certificates and private keys for
+  your self-hosted database.
```
nit: CAs only issue certificates

```suggestion
- Optional: a certificate authority that issues certificates for
  your self-hosted database.
```

GavinFrazar commented on docs/pages/database-access/guides/clickhouse-self-hosted.mdx:
```diff
@@ -53,6 +53,9 @@ choose:
 </TabItem>
 </Tabs>
 
+- Optional: a certificate authority that issues certificates and private keys for
+  your self-hosted database.
```
nit: CAs only issue certificates

```suggestion
- Optional: a certificate authority that issues certificates for
  your self-hosted database.
```

GavinFrazar commented on docs/pages/database-access/guides/cassandra-self-hosted.mdx:
```diff
@@ -25,6 +25,8 @@ description: How to configure Teleport database access with Cassandra and Scylla
 
 - Self-hosted Cassandra or ScyllaDB instance.
 - The `cqlsh` Cassandra client installed and added to your system's `PATH` environment variable.
+- Optional: a certificate authority that issues certificates and private keys for
+  your self-hosted database.
```
nit: CAs only issue certificates

```suggestion
- Optional: a certificate authority that issues certificates for
  your self-hosted database.
```

ptgott committed:
Respond to feedback

- Change `db-client` to `db_client` in commands beside `tctl auth
  export`. (greedy52)
- Use consistent Proxy Service example addresses in the Clickhouse
  guide. (Tener)
- Fix wording around the function of a CA. (GavinFrazar)
- Clarify remote `tctl` permissions in partials. (GavinFrazar)
- Clarify the need for `Db` impersonation privileges. (GavinFrazar)

ptgott commented:
Thanks for the reviews! I am responding to feedback in this PR this week. I've pushed some partial changes, and will address the rest later in the week.

> Did you test each particular engine?

@Tener I tested the self-hosted Postgres flow. For the others, I based the changes on my Postgres experience while modifying the instructions according to the requirements of each database.

ptgott commented on docs/pages/includes/database-access/tctl-auth-sign-3-files.mdx:
```diff
@@ -0,0 +1,57 @@
+verify certificates presented by the Teleport Database Service. Self-hosted
+databases also need a certificate/key pair that Teleport can verify.
+
+If you are using Teleport Cloud, your Teleport user must be allowed to
```
The guides I've changed in this PR include this partial in different places, so I'll go through and make sure the instructions are correct in each guide. I might need to refactor the partials I'm using.

GavinFrazar commented on docs/pages/database-access/guides/clickhouse-self-hosted.mdx:
```diff
@@ -121,95 +111,149 @@ you're running multiple services add `--output=stdout` to print the config in
+1. Export a CA certificate for the custom CA and make it available at `db.ca` on
+   your workstation.
+
+1. Create a secret containing the database CA certificate in the same namespace
+   as Teleport using the following command:
```
CA certificates are public info - why create a secret? (i'm not kube savvy btw)

ptgott replied:
This is based on the guidance for using `ca_cert_file` with `teleport-kube-agent` in the Helm chart reference: https://goteleport.com/docs/reference/helm-reference/teleport-kube-agent/#databases

I suppose we could also use a ConfigMap instead of a secret and create a volume from it in the same way—I think it would be functionally the same.

GavinFrazar replied:
oh ok. I think it's fine to leave it as a secret since we already do that elsewhere. I also noticed CrunchyData's PG Operator guide does the same thing for CA certs, I just didn't understand why

hugoShaka replied:
There's nothing blocking you from putting CA certs in a configmap, this is what kube automatically does for its CA cert and makes sense from a security point of view.

Putting the CA certs in a secret is a common practice in kube. I think this is because you usually also have private material (key/cert for a TLS webserver, or the CA private key when it's a private CA managed by cert-manager in the same cluster).

Another reason is that users have very limited understanding of asymmetric cryptography and often do unsafe things like sending the CA private key in a context where only the cert is needed (I've seen this many times in the wild, including in support calls with customers). I'd be in favor of keeping everything in a secret so the failure mode is not that bad.

ptgott replied:
Makes sense, thanks!

GavinFrazar commented on docs/pages/includes/database-access/self-hosted-introduction.mdx:
```diff
@@ -0,0 +1,22 @@
+
+In this guide, you will:
+
+1. Configure an {{ dbConfigure }}.
+1. Join the {{ dbName }} database to your Teleport cluster.
+1. Connect to the {{ dbName }} database via the Teleport Database Service.
```
did you intend to keep this as-is? I think at least the `Configure an` part should be changed to `Configure a`, if not the change I proposed before. See:
![image](https://github.com/gravitational/teleport/assets/16846283/93fc85c2-207e-4fc6-ac34-7779aa3a1741)

GavinFrazar commented on docs/pages/includes/database-access/tctl-auth-sign-3-files.mdx:
```diff
@@ -0,0 +1,60 @@
+Service. Self-hosted databases also need a certificate/key pair that Teleport
+can verify.
+
+By default, the Teleport Database Service trusts certificates issued by a
+certificate authority managed by the Teleport Auth Service. You can either:
+- Configure your self-hosted database to trust this CA, and instruct Teleport to
+  issue a certificate for the database to present to the Teleport Database
+  Service.
+- Configure the Database Service to trust a custom CA.
```
I'd prefer tabs here since there are two distinctly separate sets of steps here.
Not a strong preference though - it reads fine if you'd prefer to keep it.

ptgott replied:
I didn't put this in tabs because some pages that include this partial include it within a tab item, and nested tabs don't look great in our docs engine. Otherwise, tabs would make sense.

greedy52 replied:
I agree that tabs are much clearer. @kimlisa was reading this preview to fix a WebUI issue on configuring db server with custom CAs and wasn't confused after reading these paragraphs.

Any other way to organize this? 4 tabs instead of the nest tabs? Maybe an extra section on `custom CA` before troubleshooting? Or maybe a separate page on custom CA?

greedy52 replied:
BTW, in my opinion, the architecture diagram at the top should also be different for custom CA scenarios. Instead "postgresql certs", it is only database client CA for the PostgreSQL server.

ptgott replied:
Personally, I think `postgresql certs` is vague enough that it could apply to either scenario (e.g., "postgresql client certs"). This is also not ideal for the clarity of the diagram, but I think we can reserve the diagram edits for a separate project, if that makes sense!

GavinFrazar reviewed (approved):
I left a couple more suggestions, but it LGTM.
Thanks for updating these!

ptgott committed:
Update docs/pages/includes/database-access/self-hosted-introduction.mdx

Co-authored-by: Gavin Frazar <gavin.frazar@goteleport.com>

greedy52 commented on docs/pages/includes/database-access/tctl-auth-sign-3-files.mdx:
````diff
@@ -0,0 +1,60 @@
+   workstation:
+
+   ```code
+   $ tctl auth export --type=db-client --auth-server=<Var name="example.teleport.sh:443" /> > server.cas
````
this `server.cas` cannot be used by the db server directly. Instead, we should append our db client CA to their existing `server.cas`

ptgott replied:
👍 I've updated the partial. I'd also be interested in knowing why this is!

GavinFrazar replied:
It just needs to have both Teleport's CA and their custom CA, otherwise their custom CA client certs will get rejected.

Whether that be 
"append your custom CA to server.cas"
or
"append Teleport's CA to your CA file"

Both will work.

GavinFrazar replied:
Your updated version works perfectly btw 👍

greedy52 commented on docs/pages/includes/database-access/tctl-auth-sign-3-files.mdx:
```diff
@@ -0,0 +1,60 @@
+
+   The command creates 1 file, `server.cas`.
+
+1. Generate `server.crt` and `server.key` by retrieving a TLS certificate and
+   private key from your existing database CA, signed for your database server.
+   You will use these files later in the guide.
```
I feel in most scenarios that people want to use custom CAs, their PostgreSQL servers are already running with their own custom CA signed certs. In that case, we just need to append the db client CA in the previous step.

If they hasn't configured TLS on their PostgreSQL server yet, it would be easier they go option 1 to use Teleport's DB server CAs.

In case users want to use their own CA but hasn't configured the database servers yet (especially the cluster setups), @GavinFrazar once pointed out that the users may not know how to generate the self-signed certs. We might want to provide openssl instructions in that case, in the future. Or maybe a docker-compose file or how to use helm charts to deploy the cluster setup.

ptgott replied:
> I feel in most scenarios that people want to use custom CAs, their PostgreSQL servers are already running with their own custom CA signed certs. In that case, we just need to append the db client CA in the previous step.

👍 Do the changes in bba4cd523e132035bfdaa9f6ed70e433e0a223c1 take care of this?

> In case users want to use their own CA but hasn't configured the database servers yet (especially the cluster setups), @GavinFrazar once pointed out that the users may not know how to generate the self-signed certs. We might want to provide openssl instructions in that case, in the future. Or maybe a docker-compose file or how to use helm charts to deploy the cluster setup.

Since this PR has been open for a while, would it make sense to add any example instructions in a separate project?

greedy52 replied:
thanks!

ptgott committed:
Respond to greedy52 feedback

- Edit `tctl-auth-sign-3-files.mdx` to instruct the user to append the
  database client CA to your database's existing cert file.
- Use Tabs for tctl-auth-sign-3-files.mdx
